### ISSUE TITLE
Add ability to have themes inheritance and sub-themes

### ISSUE THREAD

**Issue by pascalduez** [SassDoc/sassdoc#272]
> _This is a cross issue between [SassDoc](https://github.com/SassDoc/sassdoc) and [Themeleon](https://github.com/themeleon/themeleon).
> Although most of the work would need to happens in Themeleon, I'm posting here so we can get a wider visibility_
## Observations

Nearly all custom themes we've seen released yet are forks of [default](https://github.com/SassDoc/sassdoc-theme-default) with different levels of tweaks (mostly styling). Proceeding this way makes it really hard to follow with subsequent [default](https://github.com/SassDoc/sassdoc-theme-default) updates.

We have the [generator](https://github.com/SassDoc/generator-sassdoc-theme), (which does not seem much used at the moment). It's a great tool aimed at people willing to build their custom theme from scratch (almost) and having full power on it.

We have the [blank](https://github.com/SassDoc/sassdoc-theme-blank) theme, which again promotes the fork and tweak method.
## Case study

Something really great coming from Drupal's theme system, is the [ability to define base and sub themes](https://www.drupal.org/node/225125). It allows to build an infinite inheritance tree that bubbles up to the core. A common example are base themes or starter kits, which care of a bunch of markup cleaning and structure, but leaving the finales touches to the sub-theme.

If a specific template is found/registered in a sub-theme, it will take precedence over its parent one. If not, the parent one will be used, and so on.

Assets (CSS, JS) are declared in an array/object, and can then be removed/overriden/added.

**See it as some sort of CSS cascade for templates and assets.**

The idea here is obviously not be to re-implement the whole [Drupal theme system](https://www.drupal.org/theme-guide/6-7), but get inspired and build a much lighter version.
## Suggestions

In the scope of Themeleon, and as an example SassDoc:

Theme [blank](https://github.com/SassDoc/sassdoc-theme-blank) should be renamed into `sassdoc-theme-base` (or the like) and given richer templates. It should implement the new theme system to define templates.

Theme [default](https://github.com/SassDoc/sassdoc-theme-default) should implement the new theme system to define templates and assets.

One willing to lean on core's default templates (thus benefits from updates, fixes, new features) and only do some styling could either:
- Define `sassdoc-theme-base` as base theme, register new assets.
- Define `sassdoc-theme-default` as base theme, override existing assets or register new ones.

Add an option to the [generator](https://github.com/SassDoc/generator-sassdoc-theme) to be able to use a base-theme.

That most likely means to introduce some sort of themes conventions, for instance
templates naming etc.

The `base-theme` option would be defined in the <del>`view.json`</del> `.sassdocrc` or [config object](https://github.com/SassDoc/generator-sassdoc-theme/blob/master/templates/swig/_index.js#L73-L84) of the sub-theme.

We could even extend the idea to [plugins](http://sassdoc.com/extending-sassdoc/). A plugin could provide its own set of templates and/or assets to help implement a given feature.
## Discussion
- Before going deeper into possible ways to implement this and crazy ideas/explanations,
  I would like to gather interest feedback first.
- Should we allow an infinite level of sub-theme or just two levels (base->sub) ?
- Would the level of complexity required be acceptable for Themeleon ?

**Comment by HugoGiraudel** [comment]
@valeriangalliat What is your feeling about this for 2.2?

**Comment by valeriangalliat** [comment]
Maybe 2.2, depends on how I end up implementing this.

**Comment by valeriangalliat** [comment]
Summary of a reflection on Slack, 2015-02-10, about theme inheritance.

* This should not be done inside Themeleon library; it breaks the Themeleon interface and would force theme creators to use Themeleon library (and we already don't use the library anymore in default theme).

* Something like a transparent [UnionFS](http://en.wikipedia.org/wiki/UnionFS) mount between parent theme(s) and final theme directory would be great, so a child theme can override any file without needing special support from recursive file copy algorithm, and template engines partials loader.

* Without depending on UnionFS, we can just have a `make` task that copies parent theme(s) structure inside a `dest` directory, and merge the child theme files in there to build the final tree. Basically `cp -r node_modules/parent-theme/* dest; cp -r child/* dest`. The theme would then export `dest/index.js` as package `index`.

  This way we can reuse everything from the default theme, including templates, stylesheets, and even makefile.

**Comment by HugoGiraudel** [comment]
So can we close?

**Comment by valeriangalliat** [comment]
No.

**Comment by HugoGiraudel** [comment]
How about now @valeriangalliat? 😄

**Comment by valeriangalliat** [comment]
Yeah, Ok. :D
